### PR TITLE
Search: Add inline tags

### PR DESCRIPTION
**Pull request by psealock**
The Search component sometimes requires Tags to be rendered within the input instead of below. This PR creates a prop `inlineTags` to render tags inline. 

## Screenshots

screen shot 2018-10-11 at 5 28 24 pm

## Test

1. If you have devDocs enabled, head to `/wp-admin/admin.php?page=wc-admin#/devdocs/search`. Otherwise, `/wp-admin/admin.php?page=wc-admin#/analytics/orders?filter=advanced` and add a Product filter.
2. Use the Search component to add Products
3. Is the behaviour expected? Be sure to use the keyboard to navigate as well. 

## Other Changes

1. Added a `<label/>` for the Search input.
2. Ran `npm run docs`

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -29,6 +29,9 @@
         "react/react-in-jsx-scope": 0,
         "wpcalypso/import-no-redux-combine-reducers": 0,
         "wpcalypso/jsx-classname-namespace": 0,
-        "wpcalypso/redux-no-bound-selectors": 1
+        "wpcalypso/redux-no-bound-selectors": 1,
+		"jsx-a11y/label-has-for": [ "error", {
+			"required": "id"
+		} ]
```
Comment (psealock) [inline .eslintrc]: Following Gutenberg's lead on this:
https://github.com/WordPress/gutenberg/blob/80ac62507567a7e8c2f02693bbb05edc1532407b/eslint/config.js#L44-L46

File: client/components/search/example.md [hunk 2]
```diff
@@ -4,11 +4,26 @@ import { Search } from '@woocommerce/components';
...
+		<H>Tags Below Input</H>
+		<Section component={ false }>
+			<Search
+				type="products"
+				placeholder="Search for a product"
+				selected={ selected }
+				onChange={ items => setState( { selected: items } ) }
+			/>
+		</Section>
+		<H>Tags Inline with Input</H>
```
Comment (timmyc) [inline client/components/search/example.md]: Nice! Thanks for updating the docs.

File: docs/components/filters.md [hunk 3]
```diff
@@ -85,6 +85,63 @@ The `path` parameter supplied by React-Router.
 
 The query string represented in object form.
 
+`CompareFilter` (component)
+===========================
```
Comment (timmyc) [inline docs/components/filters.md]: Thanks for re-generating the docs.

File: client/components/search/index.js [hunk 4]
```diff
@@ -71,55 +78,111 @@ class Search extends Component {
...
+									'is-active': isActive,
+								} ) }
+								onClick={ () => {
+									this.input.current.focus();
+								} }
+							>
+								{ this.renderTags() }
+								<input
+									ref={ this.input }
+									type="text"
```
Comment (timmyc) [inline client/components/search/index.js]: why is this type=text here and type=search below?

**Review by timmyc — COMMENTED** [review]
Great stuff Paul. Couple small questions in the code, but also wanted to see if @ryelle wanted to 👀 this one too due to her prior work on this component. From testing things out in Dev Docs, I noticed the page would jump around a bit as I worked with the components... typically after selecting the first product, or removing the last product:

scroll-jump

Not sure if this is due to having the two instances on the same page using the same data, but we can just watch out for this as we integrate it in reports and handle it in a future PR if it is persistent.

The last thing I noticed with the inline tags is later in the .gif above, the div appears to get quite a bit of padding on the bottom when longer product names are added. Not a massive deal but figured I'd point it out too.

File: client/components/search/index.js [hunk 5]
```diff
@@ -71,55 +78,111 @@ class Search extends Component {
...
-								<Tag
-									key={ item.id }
-									id={ item.id }
-									label={ item.label }
-									remove={ this.removeResult }
-									removeLabel={ __( 'Remove product', 'wc-admin' ) }
-									screenReaderLabel={ screenReaderLabel }
+					{ ( { listBoxId, activeId, onChange } ) =>
+						// Disable reason for no-static-element-interactions: There is no appropriate role
+						// which describes the input container intended accessible usability.
```
Comment (ryelle) [inline client/components/search/index.js]: Can you rephrase this a little? Actually you could probably merge both disable reasons into one, since it's really the same for both, right?

File: client/components/search/index.js [hunk 6]
```diff
@@ -71,55 +78,111 @@ class Search extends Component {
...
 
+	renderTags() {
+		const { selected } = this.props;
+		return selected.length ? (
+			<div className="woocommerce-search__token-list">
+				{ selected.map( ( item, i ) => {
+					const screenReaderLabel = sprintf(
+						__( '%1$s (%2$s of %3$s)', 'wc-admin' ),
+						item.label,
+						i,
```
Comment (ryelle) [inline client/components/search/index.js]: `i` should be `i+1`, the index is zero-based so we get "T-Shirt (0 of 1)"

File: client/components/search/index.js [hunk 7]
```diff
@@ -71,55 +78,111 @@ class Search extends Component {
...
+		const { placeholder, inlineTags, selected, instanceId } = this.props;
+		const { value = '', isActive } = this.state;
 		const aria = {
 			'aria-labelledby': this.props[ 'aria-labelledby' ],
 			'aria-label': this.props[ 'aria-label' ],
 		};
 		return (
 			<div className="woocommerce-search">
+				<label htmlFor={ `search-input-${ instanceId }` } className="screen-reader-text">
+					{ placeholder || __( 'Add item', 'wc-admin' ) }
```
Comment (ryelle) [inline client/components/search/index.js]: We're already using `aria-label` or `aria-labelledby` to fill this purpose, we don't actually need to add a `label` here (especially an invisible label).

File: client/components/search/index.js [hunk 8]
```diff
@@ -71,55 +78,111 @@ class Search extends Component {
...
+									id={ `search-input-${ instanceId }` }
+									value={ value }
+									placeholder={ ( selected.length === 0 && placeholder ) || '' }
+									className="woocommerce-search__inline-input"
+									onChange={ this.updateSearch( onChange ) }
+									aria-owns={ listBoxId }
+									aria-activedescendant={ activeId }
+									onFocus={ this.onFocus }
+									onBlur={ this.onBlur }
+									{ ...aria }
```
Comment (ryelle) [inline client/components/search/index.js]: It might be good to add an `aria-describedby` div that said "Move [forward/backward] for selected items" or something like that, so screen reader users know where these show up

**Review by ryelle — COMMENTED** [review]
A few notes, but looks good 👍 I don't see the jumping issue @timmyc  mentioned, though.

File: client/components/search/index.js (hunk 7 above)
Comment (psealock) [reply client/components/search/index.js]: Oh right of course. Those are passed in via props. How do you feel about making them `required`?

File: client/components/search/index.js (hunk 4 above)
Comment (psealock) [reply client/components/search/index.js]: "search" tells the browser to render the clear button on the right:

screen shot 2018-10-12 at 9 59 52 am

I don't think it makes sense in this context to have a clear button. Would it clear the text? Or all tags?

File: client/components/search/index.js (hunk 7 above)
Comment (ryelle) [reply client/components/search/index.js]: Well, you wouldn't have both– `aria-label` would just be a string, and `aria-labelledby` is an element ID. If you want to have a required prop, we could change it to accept a `label` prop, which would be a string, and we can either pass it to the input's `aria-label` or this `<label>` element. We'd have to change how the `SearchFilter` works, though (that uses `ariaLabelledby`, which is actually not right).

**Comment by psealock** [comment]
Thanks for the review @timmyc @ryelle 

* Changed `state` in the example.md so the jumpiness isn't there when looking at devDocs
* Removed the unneeded `<label>` 
* Updated the eslint disabled comment to be more readable
* Fixed item label numbering issue
* Added `aria-describedby` when items have been selected so screen readers can tell the user where selected items can be found

File: client/components/search/index.js (hunk 4 above)
Comment (timmyc) [reply client/components/search/index.js]: Ah gotcha, no I think **not** having the browser clear button makes sense for inline tags as it would likely cause confusion.

**Review by timmyc — APPROVED** [review]
Thanks for the updates and the responses on my questions. Testing well for me and looking splendid too

File: client/components/search/index.js (hunk 7 above)
Comment (psealock) [reply client/components/search/index.js]: >We'd have to change how the SearchFilter works, though (that uses ariaLabelledby, which is actually not right).

This reminds me, I've changed this in https://github.com/woocommerce/wc-admin/pull/457 to use `aria-label` as suggested
